src/System.IO.FileSystem/src/System/IO/UnixFileSystem.cs

```diff
@@ -566,12 +566,15 @@ private static string NormalizeSearchPattern(string searchPattern)
                 return searchPattern;
             }
 
-            private static Microsoft.Win32.SafeHandles.SafeDirectoryHandle OpenDirectory(string fullPath)
+            private static Microsoft.Win32.SafeHandles.SafeDirectoryHandle OpenDirectory(string fullPath, bool throwOnError = true)
             {
                 Microsoft.Win32.SafeHandles.SafeDirectoryHandle handle = Interop.Sys.OpenDir(fullPath);
                 if (handle.IsInvalid)
                 {
-                    throw Interop.GetExceptionForIoErrno(Interop.Sys.GetLastErrorInfo(), fullPath, isDirectory: true);
+                    if (throwOnError)
+                        throw Interop.GetExceptionForIoErrno(Interop.Sys.GetLastErrorInfo(), fullPath, isDirectory: true);
+                    else
+                        return null;
```

Comment:
You are correct, I had not thought about change of accesscontrol during enumeration on Windows. Using the same test but changing the Directory.Delete call to a denial of read access to the test folder results in a Exception from the enumerator.